### DISCUSSION TITLE
Accept https://crates.io/crate/<cratename> (singular)

### DISCUSSION THREAD

**Discussion by joshka** [rust-lang/crates.io#8815] [Feature Requests]
Many times when just slapping a url together I've typed /crate/some-crate instead of /crates/some-crate. I'd like to suggest that crates.io should accept this (but [mark the correct url as canonical via redirect / link rel](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Choosing_between_www_and_non-www_URLs)). This would help scenarios where the url is typed manually (e.g. in a chat / tweet, into a readme, etc.).

- docs.rs uses singular crate
- deps.rs uses singular crate
- lib.rs uses plural, but acceps singular
- There are 172 code files on github that exhibit this error https://github.com/search?q=crates.io%2Fcrate%2F&type=code

I'd guess there are probably some stats somewhere on the quantity of 404 results that hit the /crate address that would be useful to look at. I'd bet it's pretty common

**Comment by Blindspot22** [comment]
> Many times when just slapping a url together I've typed /crate/some-crate instead of /crates/some-crate. I'd like to suggest that crates.io should accept this (but [mark the correct url as canonical via redirect / link rel](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Choosing_between_www_and_non-www_URLs)). This would help scenarios where the url is typed manually (e.g. in a chat / tweet, into a readme, etc.).
> 
>     * docs.rs uses singular crate
> 
>     * deps.rs uses singular crate
> 
>     * lib.rs uses plural, but acceps singular
> 
>     * There are 172 code files on github that exhibit this error https://github.com/search?q=crates.io%2Fcrate%2F&type=code
> 
> 
> I'd guess there are probably some stats somewhere on the quantity of 404 results that hit the /crate address that would be useful to look at. I'd bet it's pretty common

Yes @joshka i think an implementation for accepting both singular and plural forms of the url reduces user frustration and errors. I have experienced it before and people often type urls manually in chats, tweets, and documentation, and having such a flexibility in url forms ensures that they reach the intended page without encountering 404 errors.

**Comment by tgross35** [comment]
It would be nice to also accept links without `crates` or `crate`; for example, redirecting crates.io/libc to crates.io/crates/libc if there is no other `libc` page on crates.io. Docs.rs does something similar, e.g. https://docs.rs/libc/.
